Support slice.Tensor for nested tensors (forward only)

jbschlosser opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #83550
* __->__ #83736

Adds NT support for `slice.Tensor` on regular dimensions. This is used in `narrow`, `chunk`, `split`, and `split_with_sizes`, so we get those for (nearly) free. Note that the optional (start, end) arg handling is factored out into a helper function `get_slice_range()` in `TensorShape.h` for use in the normal and nested impls.

TODO:
* Backwards impl
* Hook up to indexing operator for better advanced slicing support on the Python side

Differential Revision: [D39157966](https://our.internmc.facebook.com/intern/diff/D39157966)

cc @cpuhrsch @bhosmer @drisspg @mikaylagawarecki

jbschlosser committed:
Support slice, chunk, split for nested tensors

[ghstack-poisoned]

jbschlosser committed:
Update on "Support slice.Tensor for nested tensors"


Adds NT support for `slice.Tensor`. This is used in `narrow`, `chunk`, `split`, and `split_with_sizes`, so we get those for (nearly) free.

[ghstack-poisoned]

jbschlosser committed:
Update on "Support slice.Tensor for nested tensors (forward only)"


Adds NT support for `slice.Tensor` on regular dimensions. This is used in `narrow`, `chunk`, `split`, and `split_with_sizes`, so we get those for (nearly) free.

TODO:
* Backwards impl
* Hook up to indexing operator for better advanced slicing support on the Python side

[ghstack-poisoned]

drisspg commented on aten/src/ATen/native/nested/NestedTensorMath.cpp:
```diff
@@ -1180,5 +1181,52 @@ Tensor _reshape_nested(const Tensor& self, IntArrayRef proposed_shape) {
+  TORCH_CHECK(dim > 0, "Nested tensor dimension 0 cannot be sliced");
+
+  // will error if the dimension is jagged
+  auto dim_size = self.size(dim);
```
I think you could also do opt_size[dim] == -1 and then provide a more collocated error message but thats a little nitty

jbschlosser replied:
True, but the error message isn't too bad :p

drisspg commented on aten/src/ATen/native/TensorShape.h:
```diff
@@ -51,6 +51,38 @@ inline int64_t get_num_splits(const Tensor& self, int64_t split_size, int64_t di
+  int64_t end_val = end.has_value() ? end.value() : INT64_MAX;
+
+  // INT64_MAX stands for default value.
+  if (start_val == INT64_MAX) {
```
rando question is this essentially maybe_wrap_dim for start and end?

jbschlosser replied:
It's definitely clamp with start clamped to [0, dim_size] and end clamped to [start, dim_size]. Probably should be rewritten cleaner at some point :)

drisspg commented on aten/src/ATen/native/nested/NestedTensorMath.cpp:
```diff
@@ -1180,5 +1181,52 @@ Tensor _reshape_nested(const Tensor& self, IntArrayRef proposed_shape) {
+  int64_t start_val, end_val;
+  std::tie(start_val, end_val) = get_slice_range(start, end, dim_size);
+  auto len = end_val - start_val;
+  for (int64_t i = 0; i < ntensors; i++) {
```
maybe c10::irange()

jbschlosser committed:
Update on "Support slice.Tensor for nested tensors (forward only)"


Adds NT support for `slice.Tensor` on regular dimensions. This is used in `narrow`, `chunk`, `split`, and `split_with_sizes`, so we get those for (nearly) free. Note that the optional (start, end) arg handling is factored out into a helper function `get_slice_range()` in `TensorShape.h` for use in the normal and nested impls.

TODO:
* Backwards impl
* Hook up to indexing operator for better advanced slicing support on the Python side

[ghstack-poisoned]

drisspg reviewed (approved):
Looks good just really 1 small comment

albanD commented on aten/src/ATen/native/nested/NestedTensorMath.cpp:
```diff
@@ -1180,5 +1181,52 @@ Tensor _reshape_nested(const Tensor& self, IntArrayRef proposed_shape) {
+    TORCH_CHECK_INDEX(false, "slice() cannot be applied to a 0-dim tensor.");
+  }
+  dim = maybe_wrap_dim(dim, ndim);
+  TORCH_CHECK(dim > 0, "Nested tensor dimension 0 cannot be sliced");
```
Is there a fundamental reason for that? Or just that this is not implemented here?

jbschlosser replied:
There's no fundamental reason - in fact, I added an impl since it was easy to special case when dim=0.

This enables things like this:
```python
nt[:]
nt[:2]
nt[1:2]
```

Note that full advanced indexing still doesn't work since select for NT only supports dim 0. This still fails for now:
```python
nt[:, 1, 1]  # RuntimeError: NestedTensor can only be selected along dimension 0, got dimension 1 instead.
```

albanD reviewed (approved):
Nice!
We will need a custom backward that saves the input's nested size but should be fairly simple after Mikayla's change. I think it can be done in a follow up as well.

jbschlosser committed:
Update on "Support slice.Tensor for nested tensors (forward only)"


Adds NT support for `slice.Tensor` on regular dimensions. This is used in `narrow`, `chunk`, `split`, and `split_with_sizes`, so we get those for (nearly) free. Note that the optional (start, end) arg handling is factored out into a helper function `get_slice_range()` in `TensorShape.h` for use in the normal and nested impls.

TODO:
* Backwards impl
* Hook up to indexing operator for better advanced slicing support on the Python side

[ghstack-poisoned]

jbschlosser commented:
@pytorchbot merge

albanD committed:
Update on "Support slice.Tensor for nested tensors (forward only)"


Adds NT support for `slice.Tensor` on regular dimensions. This is used in `narrow`, `chunk`, `split`, and `split_with_sizes`, so we get those for (nearly) free. Note that the optional (start, end) arg handling is factored out into a helper function `get_slice_range()` in `TensorShape.h` for use in the normal and nested impls.

TODO:
* Backwards impl
* Hook up to indexing operator for better advanced slicing support on the Python side

[ghstack-poisoned]

albanD commented:
Ho well I completely missed this while doing the rebase.
I guess the two options we have here are:
- remove the native call in split_with_sizes to make this dispatch properly
- add a is_nested() check in the slice CompositeExplicit kernel to properly handle nested even though it doesn't dispatch

This first one would be cleaner but will remove the perf benefit of https://github.com/pytorch/pytorch/pull/83213 is that ok to do @swolchok ?

drisspg commented:
@swolchok re-pinging here

erichan1 commented:
@erichan1 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D39157966).

drisspg commented:
> Ho well I completely missed this while doing the rebase. I guess the two options we have here are:
> 
> * remove the native call in split_with_sizes to make this dispatch properly
> * add a is_nested() check in the slice CompositeExplicit kernel to properly handle nested even though it doesn't dispatch
> 
> This first one would be cleaner but will remove the perf benefit of #83213 is that ok to do @swolchok ?

@albanD I know Eric is looking into trying to land the top layer of this stack. Do you think we can just add a `.is_nested()` call ?

drisspg committed:
Update on "Support slice.Tensor for nested tensors (forward only)"


Adds NT support for `slice.Tensor` on regular dimensions. This is used in `narrow`, `chunk`, `split`, and `split_with_sizes`, so we get those for (nearly) free. Note that the optional (start, end) arg handling is factored out into a helper function `get_slice_range()` in `TensorShape.h` for use in the normal and nested impls.

TODO:
* Backwards impl
* Hook up to indexing operator for better advanced slicing support on the Python side

Differential Revision: [D39157966](https://our.internmc.facebook.com/intern/diff/D39157966)

[ghstack-poisoned]

drisspg committed:
Update on "Support slice.Tensor for nested tensors (forward only)"


Adds NT support for `slice.Tensor` on regular dimensions. This is used in `narrow`, `chunk`, `split`, and `split_with_sizes`, so we get those for (nearly) free. Note that the optional (start, end) arg handling is factored out into a helper function `get_slice_range()` in `TensorShape.h` for use in the normal and nested impls.

TODO:
* Backwards impl
* Hook up to indexing operator for better advanced slicing support on the Python side

Differential Revision: [D39157966](https://our.internmc.facebook.com/intern/diff/D39157966)

[ghstack-poisoned]

drisspg commented on aten/src/ATen/native/nested/NestedTensorMath.cpp:
```diff
@@ -1180,5 +1182,58 @@ Tensor _reshape_nested(const Tensor& self, IntArrayRef proposed_shape) {
+      int64_t *size_ptr = new_sizes[i].data_ptr<int64_t>();
+      int64_t *stride_ptr = new_strides[i].data_ptr<int64_t>();
+
+      new_offsets[i] = offsets[i] + start_val * stride_ptr[dim];
```
@albanD I might forward fix: #84145 and change the offsets size back to n_tensors, but I liked my code changes so want to keep those.. The changes I made here after rebasing should be valid however. The problem is that the invariant does not hold when you slice dim != 0 meaning the n+1 element doesn't maintain any semantic value either.

drisspg replied:
actually just going to update the this PR in the stack

drisspg replied:
Update. Since the last review, I added the `is_nested` path in slice and changed back the num elements in offsets from n_tensors +1 -> n_tensors

drisspg committed:
Update on "Support slice.Tensor for nested tensors (forward only)"


Adds NT support for `slice.Tensor` on regular dimensions. This is used in `narrow`, `chunk`, `split`, and `split_with_sizes`, so we get those for (nearly) free. Note that the optional (start, end) arg handling is factored out into a helper function `get_slice_range()` in `TensorShape.h` for use in the normal and nested impls.

TODO:
* Backwards impl
* Hook up to indexing operator for better advanced slicing support on the Python side

Differential Revision: [D39157966](https://our.internmc.facebook.com/intern/diff/D39157966)

[ghstack-poisoned]

drisspg committed:
Update on "Support slice.Tensor for nested tensors (forward only)"


Adds NT support for `slice.Tensor` on regular dimensions. This is used in `narrow`, `chunk`, `split`, and `split_with_sizes`, so we get those for (nearly) free. Note that the optional (start, end) arg handling is factored out into a helper function `get_slice_range()` in `TensorShape.h` for use in the normal and nested impls.

TODO:
* Backwards impl
* Hook up to indexing operator for better advanced slicing support on the Python side

Differential Revision: [D39157966](https://our.internmc.facebook.com/intern/diff/D39157966)

[ghstack-poisoned]

swolchok commented:
> This first one would be cleaner but will remove the perf benefit of https://github.com/pytorch/pytorch/pull/83213 is that ok to do @swolchok ?

if we must. I don't think it was a huge improvement

FarzanT commented:
@drisspg 

Hi, was this feature added to PyTorch 2.0? I got this error:
```
NotImplementedError: Could not run 'aten::slice.Tensor' with arguments from the 'NestedTensorCUDA' backend.
```

I feel like it was almost added but was forgotten about last October.

drisspg commented:
@FarzanT yup parts of this PR were landed. Is there a repro for what raised the NotImplementedError ?

FarzanT commented:
@drisspg Thanks!
Perhaps I'm misunderstanding what this PR is supposed to achieve. I'm on '2.0.1'. 
My use case is something like the following:
```python
a, b = torch.arange(3), torch.arange(5) + 3
nt = torch.nested.nested_tensor([a, b, a, b])

nt[0:1, ]
```
Which outputs a long stack trace, part of the first line being:
```
NotImplementedError: Could not run 'aten::slice.Tensor' with arguments from the 'NestedTensorCPU' backend. This could be because the operator doesn't exist for this backend, or was omitted during the selective/custom build process (if using custom build)...
```

Ideally, I'd like to use a boolean mask on the first dim, which would be the batch dim, and I guess it's considered "regular." Perhaps that functionality can also be added in the future, as currently they only way is to use `torch.select()` plus extra steps to achieve the same result.

drisspg commented:
Hmm yeah the error message could be better here but currently we don't allow selecting along the outer "nested" dimension.
So for instance this works:
nt[0, 1:] but not what you have. hmmmm

FarzanT commented:
@drisspg thanks for the explanation, should I open a separate issue for this then?
